wallet: fix unnamed legacy wallet migration failure

furszy opened the pull request:
Minimal fix for #34128.

The issue occurs during the migration of a legacy unnamed wallet
(the legacy "default" wallet). When the migration fails, the cleanup
logic is triggered to roll back the state, which involves erasing the
newly created descriptor wallets directories. Normally, this only
affects the parent directories of named wallets, since they each
reside in their own directories. However, because the unnamed
wallet resides directly in the top-level `/wallets/` folder, this
logic accidentally deletes the main directory.

The fix ensures that only the wallet.dat file of the unnamed wallet
is touched and restored, preserving the wallet in BDB format and
leaving the main `/wallets/` directory intact.

#### Story Line:
#32273 fixed a different set of issues and, in doing so, uncovered
this one.
Before the mentioned PR, backups were stored in the same directory
as the wallet.dat file. On a migration failure, the backup was then
copied to the top-level `/wallets/` directory. For the unnamed legacy
wallet, the wallet directory is the `/wallets/` directory, so the source
and destination paths were identical. As a result, we threw early in the
`fs::copy_file` call ([here](https://github.com/bitcoin/bitcoin/blob/29.x/src/wallet/wallet.cpp#L4572)) because the file already existed, as we
were trying to copy the file onto itself. This caused the cleanup logic
to abort early on and never reach the removal line.


#### Testing Notes:
Cherry-pick the test commit on top of master and run it. You will
see the failure and realize the reason by reading the test code.

furszy committed:
wallet: fix unnamed wallet migration failure

When migrating any legacy unnamed wallet, a failed migration would
cause the cleanup logic to remove its parent directory. Since this
type of legacy wallet lives directly in the main '/wallets/' folder,
this resulted in unintentionally erasing all wallets, including the
backup file.

To be fully safe, we will no longer call `fs::remove_all`. Instead,
we only erase the individual db files we have created, leaving
everything else intact. The created wallets parent directories are
erased only if they are empty.
As part of this last change, `RestoreWallet` was modified to allow
an existing directory as the destination, since we no longer remove
the original wallet directory (we only remove the files we created
inside it). This also fixes the restore of top-level default wallets
during failures, which were failing due to the directory existence
check that always returns true for the /wallets/ directory.

This bug started after:
https://github.com/bitcoin/bitcoin/commit/f6ee59b6e2995a3916fb4f0d4cbe15ece2054494
Previously, the `fs::copy_file` call was failing for top-level wallets,
which prevented the `fs::remove_all` call from being reached.

furszy committed:
test: add coverage for unnamed wallet migration failure

Verifies that a failed migration of the unnamed (default) wallet
does not erase the main /wallets/ directory, and also that the
backup file exists.

furszy commented:
> In #34128, the logs say that migration was successful, which is confusing as to how they ended up in the failure case. It looks like that log line is a little misleading since we can hit a failure after it is logged if the migrated wallet(s) cannot be reloaded, which will result in the failure cleanup being hit.
> 
> So a followup question is what happened that migration thinks it was successful, but subsequent loading failed.
> 
> Anyways, deleting user wallets is super bad and we should fix that ASAP.

Yeah, the failure must be related to reloading the wallet after migration, still I didn't want to delay this fix any longer.
Any other "fail to migrate" bug would be very minor and harmless compared to removing the `/wallets/` directory.

> This appears to only affect 30.x.

Thanks for checking, so we just need to back port it there.

furszy committed:
wallet: improve post-migration logging

Right now, after migration the last message users see is "migration completed",
but the migration isn't actually finished yet. We still need to load the new wallets
to ensure consistency, and if that fails, the migration will be rolled back. This
can be confusing for users.

This change logs the post-migration loading step and if a wallet fails to load and
the migration will be rolled back.

furszy committed:
wallet: migration, fix watch-only and solvables wallets names

Because the default wallet has no name, the watch-only and solvables
wallets created during migration end up having no name either.

This fixes it by applying the same prefix name we use for the backup
file for an unnamed default wallet.

Before: watch-only wallet named "_watchonly"
After:  watch-only wallet named "default_wallet_watchonly"

furszy reviewed (commented):
Added two more commits.

First one improves logging for post-migration failures, making it clearer to users (and us) if a migration is rolled back during the wallets reload process.

Second one fixes a bug where migrated watch-only and solvables wallets were incorrectly named when created from an unnamed default wallet.

pablomartin4btc reviewed (commented):
Concept ACK

Good fix! Light code review, I'll test and re-review soon...

achow101 commented:
ACK 6ecfbc7688971a506720d29fe99d1431662d2941

furszy committed:
test: coverage for migration failure when last sync is beyond prune height

achow101 commented:
ACK ca6bfda55fca7682c8f74818e7833a01d97a0ebf

darosior commented:
> whether this would significantly complicate the implementation and delay the publication of a fix?

Maybe an alternative way of getting the failsafe could be to change the location of the backup to not be the wallet's directory, which may be erased later on, but a different directory?
https://github.com/bitcoin/bitcoin/blob/755f0900a28a874a65fb843c74d8257f96b667dc/src/wallet/wallet.cpp#L4244-L4254

furszy committed:
test: restorewallet, coverage for existing dirs, unnamed wallet and prune failure

The first test verifies that restoring into an existing empty directory
or a directory with no .dat db files succeeds, while restoring into a
dir with a .dat file fails.

The second test covers restoring into the default unnamed wallet
(wallet.dat), which also implicitly exercises the recovery path used
after a failed migration.

The third test covers failure during restore on a prune node. When
the wallet last sync was beyond the pruning height.

furszy reviewed (commented):
As mentioned above, added d046da742b3b3a0db9d455250373246e52c4dd2b to further cover the recovery path through `restorewallet()`.

furszy commented:
> By allowing `RestoreWallet` to properly restore a default wallet, we accidentally introduce the wallet dir deletion bug in `RestoreWallet`.
> 
> If we are going to load a wallet after restoring (which is what normally occurs when restoring a wallet), then there is a `fs::remove_all` to remove the restored wallet if it cannot be loaded. This is the same problem as in migration for default walle

Yes. I have been working on that fix the past hour. Already nuked the `fs::remove_all()` call in `RestoreWallet()`.

achow101 commented on src/wallet/wallet.cpp:
```diff
@@ -478,10 +478,34 @@ std::shared_ptr<CWallet> RestoreWallet(WalletContext& context, const fs::path& b
+            // Any .dat file indicates an existing wallet, and restoring into this
+            // directory would overwrite it.
+            for (const auto& entry : fs::directory_iterator(wallet_path)) {
+                if (entry.path().extension() == ".dat") {
```
In d9e4593afb386268fbdd5f7fc1f409ea2253a244 "wallet: fix unnamed legacy wallet migration failure"

I think it would be better to check the filenames specifically:

```suggestion
                if (entry.path().filename() == wallet_file.filename()) {
```

This also enables this commit to restore a default wallet if the wallets directory already contains other files with the `.dat` extension.

furszy replied:
> This also enables this commit to restore a default wallet if the wallets directory already contains other files with the .dat extension.

It seems bad to store several .dat files within the same folder. What do you think about leaving it as is for now and discuss it further in a follow-up? If there is someone doing this, it will receive a clear error message on how to fix it with the current code anyway.

achow101 replied:
My concern is more to do with people who were doing the old wallet file multiwallet (instead of wallet directory multiwallet) and named those wallet files `<something>.dat`. I believe in those cases, the default wallet would not be able to be restored after migration with the current code.

furszy replied:
Pushed.

achow101 commented on test/functional/wallet_backup.py:
```diff
@@ -136,6 +136,41 @@ def restore_wallet_existent_name(self):
+        assert_raises_rpc_error(-36, error_message, node.restorewallet, wallet_name, backup_file)
+        assert wallet_dir.exists()
+        # Clean for follow-up tests
+        os.remove(node.wallets_path / wallet_name / "wallet.dat")
```
In 67051872c4c0b0f4410a4389e234352d10fa3011 "test: coverage for migration failure when last sync is beyond prune height"

CI failure here. The wallet needs to be unloaded before we delete it.

furszy committed:
wallet: RestoreWallet failure, erase only what was created

Track what RestoreWallet creates so only those files and directories
are removed during a failure and nothing else. Preexisting paths
must be left untouched.

Note:
Using fs::remove_all() instead of fs::remove() in RestoreWallet does
not cause any problems currently, but the change is necessary for the
next commit which extends RestoreWallet to work with existing directories,
which may contain files that must not be deleted.

furszy commented on test/functional/wallet_migration.py:
```diff
@@ -659,6 +660,13 @@ def test_wallet_with_path(self, wallet_path):
+        # Test cleanup: Clear unnamed default wallet for subsequent tests
+        (self.old_node.wallets_path / "wallet.dat").unlink()
+        (self.master_node.wallets_path / "wallet.dat").unlink(missing_ok=True)
+        shutil.rmtree(self.master_node.wallets_path / "default_wallet_watchonly", ignore_errors=True)
```
done as suggested

furszy commented on test/functional/wallet_migration.py:
```diff
@@ -698,6 +698,10 @@ def test_default_wallet_failure(self):
+        # Verify the original unnamed wallet was restored
+        assert (self.master_node.wallets_path / "wallet.dat").exists()
+        # And verify it is still a BDB wallet
+        self.assert_is_bdb("")
```
Yeah, it became irrelevant after the last rework -> https://github.com/bitcoin/bitcoin/pull/34156#issuecomment-3711779782. Squashed the test changes and dropped commit.

furszy commented on test/functional/wallet_backup.py:
```diff
@@ -136,6 +136,41 @@ def restore_wallet_existent_name(self):
+        self.log.info("Test restore failure due to a .dat file in the destination directory")
+        error_message = "Failed to restore wallet. Database file exists in '{}'.".format(wallet_dir)
+        assert_raises_rpc_error(-36, error_message, node.restorewallet, wallet_name, backup_file)
+        assert wallet_dir.exists()
```
Done as suggested.

furszy commented on test/functional/wallet_migration.py:
```diff
@@ -697,14 +723,15 @@ def test_default_wallet_failure(self):
         # Check backup file exists. Because the wallet has no name, the backup is prefixed with 'default_wallet'
-        assert (self.master_node.wallets_path / f"default_wallet_{mocked_time}.legacy.bak").exists()
+        backup_path = self.master_node.wallets_path / f"default_wallet_{mocked_time}.legacy.bak"
+        assert backup_path.exists()
```
Done as suggested

achow101 commented on test/functional/wallet_backup.py:
```diff
@@ -136,6 +137,49 @@ def restore_wallet_existent_name(self):
+        os.remove(wallet_file)
+        extra_file = node.wallets_path / wallet_name / "not_a_wallet.txt"
+        extra_file.touch()
+        res = node.restorewallet(wallet_name, backup_file)
```
In 69904d4ac52495127c33f0349c4fa3d3c4bef129 "test: restorewallet, coverage for existing dirs, unnamed wallet and prune failure"

This still needs to be unloaded before the wallet file is deleted at the end of the test.

w0xlt commented on src/wallet/wallet.cpp:
```diff
@@ -4320,20 +4359,36 @@ util::Result<MigrationResult> MigrateLegacyToDescriptor(std::shared_ptr<CWallet>
             for (const auto& path_to_remove : paths_to_remove) fs::remove_all(path_to_remove);
```
nit:
```suggestion
            for (const auto& path_to_remove : paths_to_remove) fs::remove(path_to_remove);
```

w0xlt commented on test/functional/wallet_migration.py:
```diff
@@ -675,6 +684,58 @@ def test_default_wallet(self):
+
+        # Create wallet directory with the watch-only name and a wallet file.
+        # Because the wallet dir exists, this will cause migration to fail.
+        watch_only_dir = self.master_node.wallets_path / "default_wallet_watchonly"
```
nit:
Commit 3 (https://github.com/bitcoin/bitcoin/pull/34156/commits/7d817d8b675e6abb5ab37ccbc75fed094bc34e4e) could use the correct path from the start.

furszy replied:
The `default` prefix? the fix for the naming issue comes later, in c8d7c2e48af9061d8d37860f56aba5dbc4465f99.

w0xlt reviewed (commented):
ACK https://github.com/bitcoin/bitcoin/pull/34156/commits/d4eca7c67ada6142c6937780b26254fc3a61919e

stickies-v commented on src/wallet/wallet.cpp:
```diff
@@ -478,13 +480,34 @@ std::shared_ptr<CWallet> RestoreWallet(WalletContext& context, const fs::path& b
+
+            // Check we are not going to overwrite an existing db file
+            if (fs::exists(wallet_file)) {
+                error = Untranslated(strprintf("Failed to restore wallet. Database file exists in '%s'.", fs::PathToString(wallet_file)));
```
nit: "Database file exists **in**" vs "Database file exists" in the above case.

furszy commented on src/wallet/wallet.cpp:
```diff
@@ -478,13 +480,34 @@ std::shared_ptr<CWallet> RestoreWallet(WalletContext& context, const fs::path& b
+        } else {
+            // The directory doesn't exist, create it
+            if (!TryCreateDirectories(wallet_path)) {
+                error = Untranslated(strprintf("Failed to restore database path '%s'.", fs::PathToString(wallet_path)));
```
will do if have to re-touch. Otherwise it is material for a small follow-up.

ryanofsky commented on src/wallet/wallet.cpp:
```diff
@@ -4341,8 +4354,17 @@ util::Result<MigrationResult> MigrateLegacyToDescriptor(std::shared_ptr<CWallet>
+                wallet_files_to_remove.insert(files.begin(), files.end());
+                if (wallet->GetName() != wallet_name) {
+                    // Only when this is not the main wallet, it is fully safe to delete the parent path
+                    // because we created the wallet during migration
```
In commit "wallet: fix unnamed legacy wallet migration failure" (239f78641256bcce6ca5db3e34c49e50af9df5ca)

I find this comment misleading because unless the wallet is a bare BDB .dat file located in the top level directory, it is completely safe to remove the parent path. In the normal case, not removing the parent path is just an implementation choice, not a safety issue.

Would suggest rephrasing "Remove the empty wallet directory if this is a watchonly or solvables wallet directory. Keep the directory otherwise to hold the restored wallet backup. (Also, the directory couldn't be removed if the wallet is an old top-level file not in its own directory.)"

ryanofsky commented on src/wallet/wallet.cpp:
```diff
@@ -478,13 +480,22 @@ std::shared_ptr<CWallet> RestoreWallet(WalletContext& context, const fs::path& b
         }
 
-        if (fs::exists(wallet_path) || !TryCreateDirectories(wallet_path)) {
+        if (fs::exists(wallet_path)) {
```
In commit "wallet: RestoreWallet failure, erase only what was created" (5ed59557b91f85d2c36d911bd2a334cb80b7494c)

This fs::exists call is redundant and could be dropped. TryCreateDirectories already returns true if the directory was created, false if it existed, and it throws an exception on other errors.

So this could simply be:

```c++
created_parent_dir = TryCreateDirectories(wallet_path);
if (!created_parent_dir) {
    error = Untranslated(strprintf("Failed to create database path '%s'. Database already exists.", fs::PathToString(wallet_path)));
    status = DatabaseStatus::FAILED_ALREADY_EXISTS;
}
```

ryanofsky commented on src/wallet/wallet.cpp:
```diff
@@ -4385,9 +4413,15 @@ util::Result<MigrationResult> MigrateLegacyToDescriptor(std::shared_ptr<CWallet>
+        // Second, delete the created wallet directories and nothing else. They must be empty at this point.
+        for (const fs::path& dir : wallet_empty_dirs_to_remove) {
+            Assert(fs::is_empty(dir));
+            fs::remove(dir);
```
In commit "wallet: fix unnamed legacy wallet migration failure" (239f78641256bcce6ca5db3e34c49e50af9df5ca)

I think commit message could be improved. It doesn't look like this bug depends on on the wallet being named or unnamed. If the wallet name is `"wallet.dat"` or is the name of any top-level database file not inside a dedicated wallet directory, it looks like this bug would still happen. So would recommend replacing "unnamed legacy wallet" with "bare wallet file" or "top-level wallet file". (For reference the change which started creating wallets as directories rather than bare files was be8ab7d082228d09ca529d1a08730d7d5aacb0ed).

Also, IMO it would be helpful to mention the commit which introduced the directory deletion bug: f6ee59b6e2995a3916fb4f0d4cbe15ece2054494, since if you just look at the remove code, you might think the bug has been around since that code was written, but actually it was introduced more recently when the backup location changed in f6ee59b6e2995a3916fb4f0d4cbe15ece2054494.

achow101 replied:
> Would also still be curious to know if backup/restore is even necessary at all, or if the "rename (not copy) the original wallet to back it up, and rename it back if migration fails" approach would work.

It is probably viable now, but was not viable when this was first implemented. When migration still required BDB to open the wallet file, it is potentially dangerous to copy just the wallet.dat file as data can be stored in the BDB log files. The use of backup was required to ensure that all data from the log files made it into the wallet.dat file. Since migration switched to the BDB parser implementation, we are assuming that the wallet.dat file has all of the data so the backup is just doing a copy of the file directly.

But we still need to do `remove()` in the cleanup since sqlite will still generate additional files that need to be removed.

furszy replied:
> It would just be nice if there is a small fix for this bug which also simplifies migration code and removes need for name.empty() special cases everywhere, and then the remove_all calls could be dealt with separately.

I understand your point. It just feels like we're entering the unknown after what happened if we only change the wallet name and leave the `remove_all` calls. I think we should take the safest approach and get rid of them. The PR shouldn’t involve too many more lines; +80% of it is about tests.

ryanofsky reviewed (approved):
Code review d4eca7c67ada6142c6937780b26254fc3a61919e. I think this is all good, but so far I only looked at the code, not the tests, and want to look more closely and make sure I understand things before adding an ack. I left a few suggestions below, mostly about improving comments and simplifying code.

furszy commented on src/wallet/wallet.cpp:
```diff
@@ -480,10 +480,22 @@ std::shared_ptr<CWallet> RestoreWallet(WalletContext& context, const fs::path& b
+            }
+
+            // Check we are not going to overwrite an existing db file
+            if (fs::exists(wallet_file)) {
```
The restore directory is always the original wallet directory. And it can contain other files:

1) if it is the unnamed wallet, we are talking about the top-level dir `/wallets/`, which can contain other wallets, either as .dat files, directories or also what I will mention next.

2) If it is a named/unnamed wallet, the directory will have the BDB files we currently don't remove. See https://github.com/bitcoin/bitcoin/pull/34156#issuecomment-3712529470.

Also, just as a note: before the last push, we had a "no other .dat files are allowed inside the restore directory" check here, but achow convinced me that this restriction would prevent migration for some wallets so I ended adding coverage for it as well (see https://github.com/bitcoin/bitcoin/pull/34156#discussion_r2662940322).

ryanofsky commented on src/wallet/wallet.cpp:
```diff
@@ -497,7 +508,13 @@ std::shared_ptr<CWallet> RestoreWallet(WalletContext& context, const fs::path& b
     // Remove created wallet path only when loading fails
     if (load_after_restore && !wallet) {
-        fs::remove_all(wallet_path);
+        if (wallet_file_copied) fs::remove(wallet_file);
```
re: https://github.com/bitcoin/bitcoin/pull/34156#discussion_r2665857516

Thanks for the explanation. I'd suggest stating the purpose of this change in the commit message to make this clear. Would suggest something like "Using fs::remove_all() instead of fs::remove() in RestoreWallet does not cause any problems currently, but is necessary for the next commit which extends RestoreWallet to work with existing directories, which may contain files that should not be deleted."

furszy commented on src/wallet/wallet.cpp:
```diff
@@ -4281,7 +4290,7 @@ util::Result<MigrationResult> MigrateLegacyToDescriptor(std::shared_ptr<CWallet>
     // the name of the data file is used, and in the case where the wallet name
     // is blank, "default_wallet" is used.
-    const std::string backup_prefix = wallet_name.empty() ? "default_wallet" : [&] {
+    const std::string backup_prefix = wallet_name.empty() ? MigrationPrefixName(*local_wallet) : [&] {
```
Remove the relative path part? I'm pretty sure that such change would fail for the relative path migration tests.

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/34156#discussion_r2666002336

> Remove the relative path part? I'm pretty sure that such change would fail for the relative path migration tests.

Oops, sorry this was a bad suggestion. I just misread the code.

furszy replied:
np, resolved.

achow101 commented:
ACK b7c34d08dd9549a95cffc6ec1ffa4bb4f81e35eb

davidgumberg commented on test/functional/wallet_backup.py:
```diff
@@ -136,6 +137,61 @@ def restore_wallet_existent_name(self):
+        # Ensure the wallet file remains untouched
+        assert wallet_dir.exists()
+        assert_equal(original_shasum, sha256sum_file(wallet_file))
+
```
non-blocking-suggestion (cont from https://github.com/bitcoin/bitcoin/pull/34156#discussion_r2666461417)

```suggestion

        self.log.info("Test restore failure due to existing db file is the destination directory")
        original_shasum = sha256sum_file(wallet_file)
        error_message = "Failed to restore wallet. Database file exists '{}'.".format(wallet_file)
        assert_raises_rpc_error(-36, error_message, node.restorewallet, f"{wallet_name}/wallet.dat", backup_file)
        # Ensure the wallet file remains untouched
        assert wallet_dir.exists()
        assert_equal(original_shasum, sha256sum_file(wallet_file))

```

w0xlt reviewed (commented):
ACK https://github.com/bitcoin/bitcoin/commit/b7c34d08dd9549a95cffc6ec1ffa4bb4f81e35eb

davidgumberg commented:
crACK https://github.com/bitcoin/bitcoin/commit/b7c34d08dd9549a95cffc6ec1ffa4bb4f81e35eb

Did some trivial manual testing that resembles the functional tests added here, and I could produce the issue in master, 30.0, but not in 29.0 or in this branch. I believe this fixes #31428. 

Switching all `remove_all` calls to instead just `remove()` empty directories (which fails to delete directories if they're not empty) and specific wallet.dat files corresponding with `CWallet` objects we've created related to migration makes me much more confident that there isn't a way here for user data to get deleted, but I strongly agree with @willcl-ark's suggestion [above](https://github.com/bitcoin/bitcoin/pull/34156#issuecomment-3710047237) that we should **never** delete any migration data, or IMO wallet data generally, even if we are pretty confident that the data is not useful or interesting to the user, but to be clear, I think this PR works as a minimal fix.

This doesn't rise to the same level of criticality, but grepping for `remove_all` reveals that the wallet tool has a similar issue where it will delete the wallet directory on failure to load from a dump:

```bash
# !!!!CAUTION!!! DON'T RUN THIS ON A MACHINE WITH ACTUAL WALLETS, IF THE DATADIR IS NOT SET CORRECTLY, THIS MAY DELETE YOUR WALLET AND DATADIR
mkdir -p /tmp/tool/wallets
# create a wallet
./build/bin/bitcoin-wallet -datadir=/tmp/tool -wallet="example" create
# dump the wallet
./build/bin/bitcoin-wallet -datadir=/tmp/tool -wallet="example" -dumpfile="test.dump" dump
# malform the dump checksum by appending an `a`
sed -i '$ s/$/a/' test.dump
# create an unnamed wallet from a dump
./build/bin/bitcoin-wallet -datadir=/tmp/tool -dumpfile="test.dump" createfromdump
ls /tmp/tool
# wallets dir is gone
```

achow101 commented:
> This doesn't rise to the same level of criticality, but grepping for `remove_all` reveals that the wallet tool has a similar issue where it will delete the wallet directory on failure to load from a dump:

Oof yeah. Opened #34215 to fix that.

alswaryaldwly-wq commented:
Please contact me via WhatsApp 00967782452027

في الأربعاء، ٧ يناير ٢٠٢٦, ٥:٤٧ ص Matias Furszyfer ***@***.***>
كتب:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In src/wallet/wallet.cpp
> <https://github.com/bitcoin/bitcoin/pull/34156#discussion_r2666874759>:
>
> > @@ -4385,9 +4413,15 @@ util::Result<MigrationResult> MigrateLegacyToDescriptor(std::shared_ptr<CWallet>
>              }
>          }
>
> -        // Delete the wallet directories
> -        for (const fs::path& dir : wallet_dirs) {
> -            fs::remove_all(dir);
> +        // First, delete the db files we have created throughout this process and nothing else
> +        for (const fs::path& file : wallet_files_to_remove) {
> +            fs::remove(file);
> +        }
> +
> +        // Second, delete the created wallet directories and nothing else. They must be empty at this point.
> +        for (const fs::path& dir : wallet_empty_dirs_to_remove) {
> +            Assert(fs::is_empty(dir));
> +            fs::remove(dir);
>
> ah, that's much clear now. Thanks.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/bitcoin/bitcoin/pull/34156#discussion_r2666874759>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/B3ABZXLCML3NN6UAD6T3OXT4FRXUJAVCNFSM6AAAAACQDD7FOKVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZTMMZTGE2TKNJSG4>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>

willcl-ark reviewed (approved):
ACK b7c34d08dd9549a95cffc6ec1ffa4bb4f81e35eb

Happy with the fix here for the immediate issue. I think we can do even better longer-term and #34193 has a few ideas I'd like to see make it in, such as resolving the race between bdb and sqlite database files (delete->creation).

Also verified that this is not present on 29.x by backporting the tests, where we see unnamed wallet migration fail at the file copy, and so exit without the problematic removal:

<details>
<summary>29.x log</summary>

```code
2026-01-07T10:31:08.168000Z TestFramework (INFO): Test failure during unnamed (default) wallet migration
2026-01-07T10:31:08.437000Z TestFramework (ERROR): Unexpected exception
Traceback (most recent call last):
  File "/home/will/src/core/backports/test/functional/test_framework/util.py", line 160, in try_rpc
    fun(*args, **kwds)
    ~~~^^^^^^^^^^^^^^^
  File "/home/will/src/core/backports/./build/test/functional/wallet_migration.py", line 117, in migrate_and_get_rpc
    migrate_info = self.master_node.migratewallet(wallet_name=wallet_name, **kwargs)
  File "/home/will/src/core/backports/test/functional/test_framework/coverage.py", line 50, in __call__
    return_val = self.auth_service_proxy_instance.__call__(*args, **kwargs)
  File "/home/will/src/core/backports/test/functional/test_framework/authproxy.py", line 151, in __call__
    raise JSONRPCException(response['error'], status)
test_framework.authproxy.JSONRPCException: filesystem error: cannot copy file: File exists [/tmp/bitcoin_func_test_yg5hrc0g/node0/regtest/wallets/default_wallet_1767781868.legacy.bak] [/tmp/bitcoin_func_test_yg5hrc0g/node0/regtest/wallets/default_wallet_1767781868.legacy.bak] (-1)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/will/src/core/backports/test/functional/test_framework/test_framework.py", line 135, in main
    self.run_test()
    ~~~~~~~~~~~~~^^
  File "/home/will/src/core/backports/./build/test/functional/wallet_migration.py", line 1408, in run_test
    self.test_default_wallet_failure()
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^
  File "/home/will/src/core/backports/./build/test/functional/wallet_migration.py", line 568, in test_default_wallet_failure
    assert_raises_rpc_error(-4, "Failed to create database", self.migrate_and_get_rpc, "")
    ~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/will/src/core/backports/test/functional/test_framework/util.py", line 151, in assert_raises_rpc_error
    assert try_rpc(code, message, fun, *args, **kwds), "No exception raised"
           ~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/will/src/core/backports/test/functional/test_framework/util.py", line 164, in try_rpc
    raise AssertionError("Unexpected JSONRPC error code %i" % e.error["code"])
AssertionError: Unexpected JSONRPC error code -1
2026-01-07T10:31:08.489000Z TestFramework (INFO): Not stopping nodes as test failed. The dangling processes will be cleaned up later.
```

which leaves the wallet dir intact on failure:

```code
❯ eza -al /tmp/bitcoin_func_test_yg5hrc0g/node0/regtest/wallets/
drwxr-xr-x   - will  7 Jan 10:31 _watchonly
drwx------   - will  7 Jan 10:31 basic0
drwx------   - will  7 Jan 10:31 basic1
drwx------   - will  7 Jan 10:31 basic2
drwx------   - will  7 Jan 10:31 default_wallet
.rw-r--r-- 16k will  7 Jan 10:31 default_wallet_1767781868.legacy.bak
drwx------   - will  7 Jan 10:31 encrypted
drwx------   - will  7 Jan 10:31 imports0
drwx------   - will  7 Jan 10:31 imports0_watchonly
drwx------   - will  7 Jan 10:31 multisig0
drwx------   - will  7 Jan 10:31 multisig1
drwx------   - will  7 Jan 10:31 multisig1_solvables
drwx------   - will  7 Jan 10:31 multisig1_watchonly
drwx------   - will  7 Jan 10:31 pkcb
.rw------- 12k will  7 Jan 10:31 wallet.dat
```

</details>

pablomartin4btc reviewed (approved):
tACK b7c34d08dd9549a95cffc6ec1ffa4bb4f81e35eb

<details>
<summary>Verified <code>test_default_wallet_failure</code> fails on master (<code>/wallets/</code> folder was deleted).</summary>

```
2026-01-06T17:40:58.759997Z TestFramework (INFO): Test failure during unnamed (default) wallet migration
2026-01-06T17:40:58.960588Z TestFramework (ERROR): Unexpected exception
Traceback (most recent call last):
  File "/home/pablo/workspace/bitcoin/test/functional/test_framework/test_framework.py", line 142, in main
    self.run_test()
  File "/home/pablo/workspace/bitcoin/./build/test/functional/wallet_migration.py", line 1595, in run_test
    self.test_default_wallet_failure()
  File "/home/pablo/workspace/bitcoin/./build/test/functional/wallet_migration.py", line 698, in test_default_wallet_failure
    assert self.master_node.wallets_path.exists()
AssertionError
```

</details>

darosior commented on src/wallet/wallet.cpp:
```diff
@@ -4322,27 +4334,44 @@ util::Result<MigrationResult> MigrateLegacyToDescriptor(std::shared_ptr<CWallet>
+    auto track_for_cleanup = [&](const CWallet& wallet) {
+        const auto files = wallet.GetDatabase().Files();
+        wallet_files_to_remove.insert(files.begin(), files.end());
+        if (wallet.GetName() != wallet_name) {
```
Note `GetName()`'s doc string is now somewhat dangerously confusing: it's documented as a name only used for logging/debugging purposes, but is now relied upon for a funds-critical operation.

furszy replied:
Not "now". This has been the case since a long time ago. Migration was introduced relying on it. See how we create the sqlite db [here](https://github.com/bitcoin/bitcoin/blob/cd6e4c9235f763b8077cece69c2e3b2025cc8d0f/src/wallet/wallet.cpp#L3830). We use the wallet's name.

drfernandogomesreal-cpu reviewed (commented):
My v
